folding-schemes/src/utils/gadgets.rs

```diff
@@ -107,3 +107,35 @@ impl<F: PrimeField> MatrixGadget<FpVar<F>> for SparseMatrixVar<F, F, FpVar<F>> {
             .collect())
     }
 }
+
+/// Interprets the given vector v as the evaluations of a dense multilinear extension of n_vars,
+/// and evaluates it at the given point. This method mimics the behaviour of
+/// `utils/mle.rs#dense_vec_to_dense_mle` + `DenseMultilinearExtension::evaluate` but in R1CS
+/// constraints, since dense multilinear extensions are not supported in ark_r1cs_std.
+pub fn eval_mle<F: PrimeField>(
+    // n_vars indicates the number of variables in the MLE
+    n_vars: usize,
+    // v is the vector of the evaluations of the dense multilinear extension (MLE)
+    v: Vec<FpVar<F>>,
+    // point is the point at which we want to evaluate the MLE
+    point: Vec<FpVar<F>>,
+) -> FpVar<F> {
+    // pad to 2^n_vars
+    let mut poly: Vec<FpVar<F>> = [
+        v.to_owned(),
+        std::iter::repeat(FpVar::zero())
+            .take((1 << n_vars) - v.len())
+            .collect(),
+    ]
+    .concat();
+
+    for i in 1..n_vars + 1 {
+        let r = point[i - 1].clone();
+        for b in 0..(1 << (n_vars - 1)) {
+            let left = poly[b << 1].clone();
+            let right = poly[(b << 1) + 1].clone();
+            poly[b] = left.clone() + r.clone() * (right - left);
+        }
+    }
+    poly[0].clone()
```

Comment:
If we're always going to return `poly[0].clone()`, I wonder why do we need to iterate past `0th` step. In the sense, iterations after `0` will not affect the polynomial at `0`.
I assume I'm missing something. But unsure what it is.

Reply:
well, this after a double loop (one loop inside another one), where the `poly[0]` is modified multiple times and it depends on other values in the `poly` array. If we return after the 0-th iteration we will return a wrong result.

Reply:
Ohhh 
```rust
let left = poly[b << 1].clone();
let right = poly[(b << 1) + 1].clone();
``` 
Is indeed what forces to go through it all. SOrry.. I got confused.